crypto/verification.go

```diff
@@ -0,0 +1,589 @@
+		mach.Log.Error("Could not find device %v of user %v", content.FromDevice, userID)
+		return
+	}
+	// check if SAS is in supported methods
+	supportsSAS := false
+	for _, method := range content.Methods {
+		if method == event.VerificationMethodSAS {
+			supportsSAS = true
+			break
+		}
+	}
+	if !supportsSAS {
+		mach.Log.Warn("Canceling verification transaction %v as SAS is not supported", content.TransactionID)
+		mach.CancelSASVerification(otherDevice.UserID, otherDevice.DeviceID, content.TransactionID, "Only SAS method is supported", event.VerificationCancelUnknownMethod)
+		return
+	}
+	if mach.AcceptVerificationFrom(otherDevice) {
+		mach.Log.Debug("Accepting SAS verification %v from %v of user %v", content.TransactionID, otherDevice.DeviceID, otherDevice.UserID)
+		if err := mach.NewSASVerificationWith(otherDevice, content.TransactionID, 10*time.Minute, true); err != nil {
+			mach.Log.Error("Error accepting SAS verification request: %v", err)
+		}
+	} else {
+		mach.Log.Debug("Not accepting SAS verification %v from %v of user %v", content.TransactionID, otherDevice.DeviceID, otherDevice.UserID)
+		mach.CancelSASVerification(otherDevice.UserID, otherDevice.DeviceID, content.TransactionID, "Not accepted by user", event.VerificationCancelByUser)
+	}
+}
+
+// NewSASVerificationWith starts the SAS verification process with another device.
+// If the transaction ID is empty, a new one is generated.
+func (mach *OlmMachine) NewSASVerificationWith(device *DeviceIdentity, transactionID string, timeout time.Duration, useEmojiSAS bool) error {
+	if transactionID == "" {
+		transactionID = strconv.Itoa(rand.Int())
+	}
+	mach.Log.Debug("Starting new verification transaction %v with device %v of user %v", transactionID, device.DeviceID, device.UserID)
+
+	verState := &verificationState{
+		sas:                 olm.NewSAS(),
+		otherDevice:         device,
+		supportEmojiSAS:     useEmojiSAS,
+		initiatedByUs:       true,
+		verificationStarted: false,
+		keyReceived:         false,
+		sasMatched:          false,
+	}
+
+	startEvent, err := mach.StartSASVerification(device.UserID, device.DeviceID, transactionID, useEmojiSAS)
+	if err != nil {
+		return err
+	}
+
+	json, err := json.Marshal(startEvent)
+	if err != nil {
+		return err
+	}
+	canonical, err := canonicaljson.CanonicalJSON(json)
+	if err != nil {
+		return err
+	}
+
+	verState.startEventCanonical = string(canonical)
+	_, loaded := mach.keyVerificationTransactionState.LoadOrStore(device.UserID.String()+":"+transactionID, verState)
+	if loaded {
+		return errors.New("Transaction already exists")
+	}
+
+	mach.Log.Error("emoji %v", useEmojiSAS)
+
+	go mach.timeoutAfter(device.UserID, transactionID, timeout)
+
+	return nil
+}
+
+// StartSASVerification sends the SAS verification start message to another device.
+func (mach *OlmMachine) StartSASVerification(toUserID id.UserID, toDeviceID id.DeviceID, transactionID string, useEmoji bool) (*event.VerificationStartEventContent, error) {
```

Comment:
Does this need to be exported? i.e. is there any case where someone would want to use this over `NewSASVerificationWith`?

Reply:
`CancelSASVerification` seems to be similar (might not need to be exported)

Reply:
I was thinking that these could be used if someone wants to provide their own implementation for handling SAS to save them some effort, as the current one is coupled to the default syncer implementation (the handle* methods are unexported). The exported methods don't make any reference to the `verificationState` struct either, which is used by the handle* methods.

Reply:
Renamed it to `SendSASVerificationStart` / `SendSASVerificationCancel` etc and made the comment a bit more explicit.

Reply:
Looks like you renamed `StartSASVerification` to `SendSASVerificationStart`, which is better. `CancelSASVerification` should probably be renamed to `SendSASVerificationCancel`, so that the manual methods all start with `SendSASVerification`.

Related to canceling, a separate method for canceling a verification started with the `New` method might be useful.

Reply:
I thought I also renamed `CancelSASVerification`, huh. Done now.

Do you think doing user cancellation here using a context is a good idea? I'm thinking we could also `WithTimeout` on top of the given context instead of `time.Sleep(timeout) and then cancel transaction`. Then we could also have a `WithCancel` so whenever the transaction progresses, we cancel the current timeout context and start a new one with the same timeout. Or am I complicating it too much?

Reply:
I didn't like how the context thing was turning out, so I added a new `CancelSASVerification` instead for cancelling the transaction by the user.